### PR TITLE
migrate experimental-stop-grpc-service-on-defrag flag to feature gate.

### PR DESCRIPTION
**Pull request by siyuanfoundation**
Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

part of https://github.com/etcd-io/etcd/issues/18023, an example of basic wiring from an existing experimental feature flag to the new feature gate.

final step of of breaking up https://github.com/etcd-io/etcd/pull/18234/ into smaller PRs.

### CODE REVIEW

File: server/embed/config.go [hunk 1]
```diff
@@ -888,6 +888,36 @@ func (cfg *configYAML) configFromFile(path string) error {
 	return cfg.Validate()
 }
 
+// SetFeatureGatesFromExperimentalFlags sets the feature gate values if the feature gate is not explicitly set
+// while their corresponding experimental flags are explicitly set.
+// TODO: remove after all experimental flags are deprecated.
+func SetFeatureGatesFromExperimentalFlags(fg featuregate.FeatureGate, getExperimentalFlagVal func(string) (bool, bool), featureGatesFlagName, featureGatesVal string) error {
```
Comment (serathius) [inline server/embed/config.go:894]: Mark as deprecated?

File: server/etcdmain/help.go [hunk 2]
```diff
@@ -312,7 +312,7 @@ Experimental feature:
   --experimental-snapshot-catchup-entries
     Number of entries for a slow follower to catch up after compacting the raft storage entries.
   --experimental-stop-grpc-service-on-defrag
-    Enable etcd gRPC service to stop serving client requests on defragmentation.
+    Enable etcd gRPC service to stop serving client requests on defragmentation. TO BE DEPRECATED, use '--feature-gates=StopGRPCServiceOnDefrag=true' instead.
```
Comment (serathius) [inline server/etcdmain/help.go:315]: To be deprecated when? and removed when?

File: server/etcdmain/config.go [hunk 3]
```diff
@@ -238,6 +239,30 @@ func (cfg *config) configFromCmdLine() error {
 		cfg.ec.InitialCluster = ""
 	}
 
+	// getBoolFlagVal checks if the a given flag is explicitly set in the cmd line arguments.
+	getBoolFlagVal := func(flagName string) (bool, bool) {
```
Comment (serathius) [inline server/etcdmain/config.go:243]:
```suggestion
	getBoolFlagVal := func(flagName string) *bool {
```

Three value logic (true, false, unset) here would better map to pointer instead of two unnamed bool

Comment (serathius) [inline server/etcdmain/config.go:243]: Why not normal function?

File: server/config/config.go [hunk 4]
```diff
@@ -193,9 +193,6 @@ type ServerConfig struct {
 	// a shared buffer in its readonly check operations.
 	ExperimentalTxnModeWriteWithSharedBuffer bool `json:"experimental-txn-mode-write-with-shared-buffer"`
 
-	// ExperimentalStopGRPCServiceOnDefrag enables etcd gRPC service to stop serving client requests on defragmentation.
-	ExperimentalStopGRPCServiceOnDefrag bool `json:"experimental-stop-grpc-service-on-defrag"`
```
Comment (serathius) [inline server/config/config.go:197]: Would removing this be a breaking change for json config or embeded cluster? (just checking).

File: server/etcdmain/config_test.go [hunk 5]
```diff
@@ -412,16 +412,60 @@ func TestParseFeatureGateFlags(t *testing.T) {
...
+			name: "can set feature gate to true from feature gate flag",
+			args: []string{
+				"--feature-gates=StopGRPCServiceOnDefrag=true",
+			},
+			expectedFeatures: map[featuregate.Feature]bool{
 				features.StopGRPCServiceOnDefrag: true,
 			},
 		},
+		{
+			name: "can set feature gate to false from feature gate flag",
```
[the remark this answers is not in the document]
Comment (siyuanfoundation) [reply server/etcdmain/config_test.go:469]: done

File: server/embed/config.go (hunk 1 above)
Comment (siyuanfoundation) [reply server/embed/config.go]: Marking `ExperimentalFlagToFeatureMap` as deprecated, because that's where the actual experimental features are added.

File: server/etcdmain/config.go [hunk 6]
```diff
@@ -238,14 +244,44 @@ func (cfg *config) configFromCmdLine() error {
...
+	})
+	if !explicitlySet {
+		return nil
+	}
+	flagVal, parseErr := strconv.ParseBool(cfg.cf.flagSet.Lookup(flagName).Value.String())
+	if parseErr != nil {
+		panic(parseErr)
+	}
+	return &flagVal
+}
```
Comment (ahrtr) [inline server/etcdmain/config.go:274]: We can move this into https://github.com/etcd-io/etcd/blob/main/pkg/flags/flag.go as a generic function something like below,

```
func GetBoolFlagVal(fs *flag.FlagSet, flagName string) *bool {
    ......
}
```

File: server/etcdmain/config.go [hunk 7]
```diff
@@ -76,8 +78,12 @@ type configFlags struct {
 }
 
 func newConfig() *config {
+	return newConfigFromEmbedConfig(embed.NewConfig())
+}
+
+func newConfigFromEmbedConfig(ec *embed.Config) *config {
```
Comment (ahrtr) [inline server/etcdmain/config.go:84]: The implementation of `newConfig` is simple enough, let's try not to do such unnecessary change.

File: server/etcdmain/config.go [hunk 8]
```diff
@@ -238,14 +244,44 @@ func (cfg *config) configFromCmdLine() error {
...
+	}
+	return &flagVal
+}
+
 func (cfg *config) configFromFile(path string) error {
 	eCfg, err := embed.ConfigFromFile(path)
 	if err != nil {
 		return err
 	}
+	// TODO: remove after adding feature-gates field in the config file.
```
Comment (ahrtr) [inline server/etcdmain/config.go:281]: This was missed in your previous PR, I would suggest to get this resolved firstly. 

This PR is already a followup PR, let's try not add too much followup PR in the chain.

Comment (siyuanfoundation) [reply server/etcdmain/config.go:281]: ok. Let's do https://github.com/etcd-io/etcd/pull/18365 first.

File: server/etcdmain/config.go (hunk 6 above)
Comment (siyuanfoundation) [reply server/etcdmain/config.go]: done.

File: server/etcdmain/config.go (hunk 7 above)
Comment (siyuanfoundation) [reply server/etcdmain/config.go]: done.

File: server/embed/config.go [hunk 9]
```diff
@@ -897,6 +916,36 @@ func (cfg *configYAML) configFromFile(path string) error {
 	return cfg.Validate()
 }
 
+// SetFeatureGatesFromExperimentalFlags sets the feature gate values if the feature gate is not explicitly set
+// while their corresponding experimental flags are explicitly set, for all the features in ExperimentalFlagToFeatureMap.
+// TODO: remove after all experimental flags are deprecated.
+func SetFeatureGatesFromExperimentalFlags(fg featuregate.FeatureGate, getExperimentalFlagVal func(string) *bool, featureGatesFlagName, featureGatesVal string) error {
```
Comment (ahrtr) [inline server/embed/config.go:922]: You should be able to read `embed.ServerFeatureGateFlagName` directly.
```suggestion
func SetFeatureGatesFromExperimentalFlags(fg featuregate.FeatureGate, getExperimentalFlagVal func(string) *bool, featureGatesVal string) error {
```

Comment (siyuanfoundation) [reply server/embed/config.go:922]: this is passed as a param because we might want to reuse this function for `cluster-feature-gates`.

Comment (ahrtr) [reply server/embed/config.go:922]: We haven't finalized the cluster-level feature yet. We can make change later once it's finalized.

**Comment by ahrtr** [comment]
cc @fuweid @ivanvc  @jmhbnz @serathius

File: server/embed/config.go (hunk 9 above)
Comment (siyuanfoundation) [reply server/embed/config.go]: ok, removed.

**Review by fuweid — APPROVED** [review]
LGTM

File: server/embed/config_test.go [hunk 10]
```diff
@@ -743,3 +778,99 @@ func TestUndefinedAutoCompactionModeValidate(t *testing.T) {
...
+			err = SetFeatureGatesFromExperimentalFlags(fg, getExperimentalFlagVal, tc.featureGatesFlag)
+			if tc.expectErr {
+				if err == nil {
+					t.Fatal("expect error")
+				}
+				return
+			}
+			if err != nil {
+				t.Fatal(err)
+			}
```
Comment (ahrtr) [inline server/embed/config_test.go:868]:
```suggestion			
			if tc.experimentalStopGRPCServiceOnDefrag != "" {
				// overwrite the value for `features.StopGRPCServiceOnDefrag`
				// based on flag `--experimental-stop-grpc-service-on-defrag`.
				getExperimentalFlagVal := func(flagName string) *bool {
					if flagName != "experimental-stop-grpc-service-on-defrag" {
						return nil
					}
					flagVal, parseErr := strconv.ParseBool(tc.experimentalStopGRPCServiceOnDefrag)
					require.NoError(t, parseErr)
					return &flagVal
				}

				err = SetFeatureGatesFromExperimentalFlags(fg, getExperimentalFlagVal, tc.featureGatesFlag)
				if tc.expectErr {
					require.True(t, err != nil, "expect error")
					return
				}
				require.NoError(t, err)
			}			
```

Comment (ahrtr) [reply server/embed/config_test.go:868]: > This misses out the test case when the flag is not set.

You can add a separate case for that. Always remember to improve the readability and keep the test case easy to understand

Comment (siyuanfoundation) [reply server/embed/config_test.go]: personally I don't think putting the function this test is mainly about inside a if block is improving the readability. 
Added some comments, hopefully that helps.

Comment (ahrtr) [reply server/embed/config_test.go]: 1. It isn't about a if block, it's about not coupling a test driving condition `tc.experimentalStopGRPCServiceOnDefrag != ""` with a hard code condition `flagName != "experimental-stop-grpc-service-on-defrag"`. Putting the former in a outer / upper layer gives a clear hint what it's for, so easier to understand.
2. The special missed test case you mentioned above deserves a dedicated test case; do not implicitly test it in this test, otherwise other contributors may update this test in future and miss it again.
3. Use require.NoError to get rid of too many `if err != nil` or `if err == nil`

Comment (siyuanfoundation) [reply server/embed/config_test.go]: updated code to `require.NoError`

**Review by ahrtr — APPROVED** [review]
LGTM

Thanks

**Comment by ahrtr** [comment]
cc  @jmhbnz @serathius

**Review by serathius — APPROVED** [review]
